### ISSUE TITLE
MenuBar menu items do not display shortcut key on macOS with prefer_global_menu

### ISSUE THREAD

**Issue by lostminds** [godotengine/godot#82854]
### Godot version

4.2.dev6

### System information

macOS 13.6.0

### Issue description

The MenuBar control can be used with the prefer_global_menu property (default true) to setup the application top menu bar when available (on macOS only?). This generates menus and items based on the PopupMenu nodes placed in the MenuBar.

While you can't access this via the editor [I've now learnt](https://github.com/godotengine/godot-proposals/issues/7988#issuecomment-1748909861) that it is possible to set shortcuts for these menu items using script. This works  as intended for embedded menu items as shown below (apart from a bug that appends "(Unicode)"?):

![Screenshot 2023-10-05 at 16 02 06](https://github.com/godotengine/godot/assets/17763524/68af4f59-298a-41e4-918e-22f751b589b0)

However, if you toggle prefer_global_menu on again and the menu is converted to the OS standard menu these items do not get the assigned shortcut keys:
![Screenshot 2023-10-05 at 16 02 29](https://github.com/godotengine/godot/assets/17763524/e96c10b2-6ce9-438e-806c-1475dd9cb131)

For reference how macOS menu items with shortcuts normally appear:
![Screenshot 2023-10-05 at 16 12 59](https://github.com/godotengine/godot/assets/17763524/011cf465-9c83-4e4a-89b3-65a968fff3a2)

 Likely when setting up the corresponding NSMenuItems for the macOS menu, the [keyEquivalent](https://developer.apple.com/documentation/appkit/nsmenuitem/1514842-keyequivalent?language=objc) property is not set to match the Shortcut for the item.

Correctly setting this keyEquivalent would also have the benefit of the menu item automatically intercepting the key event and highlighting the menu as appropriate when the item is triggered, as well as sending the event that the item was selected.

### Steps to reproduce

Set up a project with a Control scene with:
- MenuBar
-  - PopupMenu

Configure the popup menu to have at least one item.

In ready, assign the keyboard shortcut to the menu item (in a script)
```
var menu:PopupMenu = $MenuBar/PopupMenu
var shortcut:Shortcut = Shortcut.new()
var shortcutKeyEvent:InputEventKey = InputEventKey.new()
shortcutKeyEvent.key_label = KEY_O
shortcutKeyEvent.command_or_control_autoremap = true
shortcut.events.append(shortcutKeyEvent)
menu.set_item_shortcut(0,shortcut,true)
```

Run on macOS and observer that the shortcut key is missing in the menu item.

### Minimal reproduction project

[NativeMenuItemShortcuts.zip](https://www.lostminds.com/godot/NativeMenuItemShortcuts.zip)

**Comment by bruvzg** [comment]
It's using `keycode` from the shortcut to set native menu shortcuts, not a `key_label`. But I guess should be checking for all options instead.

**Comment by lostminds** [comment]
> It's using `keycode` from the shortcut to set native menu shortcuts, not a `key_label`. But I guess should be checking for all options instead.

Ah, I understand. Thaks for the quick reply! I can confirm that using `keycode` shows the shortcuts correctly.

![Screenshot 2023-10-05 at 17 10 03](https://github.com/godotengine/godot/assets/17763524/c2632555-931a-409d-9313-eb1a1ad7fd77)

I did however find another bug: The menu shortcuts only work if you use them with the window of the original MenuBar control active. If you have another window active the menu bar item flashes showing it has received the shortcut key event and triggered, but the corresponding `index_pressed` / `id_pressed` signal isn't sent. So I'm guessing you don't use the events from the menu item and instead rely on some other way of catching these shortcut events that doesn't work if the window is not active? (even though the set_item_shortcut global flag is set to true)

**Comment by bruvzg** [comment]
> So I'm guessing you don't use the events from the menu item and instead rely on some other way of catching these shortcut events that doesn't work if the window is not active?

Menu shortcuts are redirected to the generic Godot event processing (at least at some parts of the editor it's expected behavior, and some UI subelements override menu shortcuts). There's a DisplayServer method to connect callback directly to the native menu item, but it's currently not used by the PopupMenu (I'll probably will add a flag to allow using it).

**Comment by lostminds** [comment]
How do you handle the action that comes from the NSMenuItem/NSMenu when the user uses the mouse to select a menu item? That still works for me regardless of what window is active and I'd expect the same action to be sent to the same target if the NSMenuItem is triggered by a keyEquivalent being pressed. The highlight/blink of the menu seems to indicate that the key event is causing the NSMenuItem to be triggered, so at least the key event is reaching the menu. 

I noticed what could be a clue to why selecting the items with mouse works: If I open the menu and keep it open, the menu keyboard shortcuts trigger the expected action, even if a different window is active. Could it be that the menu popup is technically a little window, and the display server is allowing the action then since it comes from an active window? This would be the same when you select a menu item with the mouse, but not when the item is triggered by shortcut in a closed menu.

**Comment by bruvzg** [comment]
I noticed what could be a clue to why selecting the items with mouse works: 

Clicking is always trigger the callback directly (`[action](action:@selector(globalMenuCallback:))` of the item):

https://github.com/godotengine/godot/blob/9e455f424a635549b295272d82757c7ff99f33d7/platform/macos/display_server_macos.mm#L867

`keyEquivalent` is handled separately and either call the action (if a dedicated key_callback is set, which PopupMenu do not do), or forwarded to the key window. This is done to make sure editor shortcuts are handled the same way on all platforms (e.g, code editor have its own built-in menu with own Undo/Redo which should trigger first).

https://github.com/godotengine/godot/blob/9e455f424a635549b295272d82757c7ff99f33d7/platform/macos/godot_menu_delegate.mm#L76-L84

**Comment by lostminds** [comment]
If I understand it correctly,` if (value->key_callback != Callable())` is where you check if this GodotMenuItem has something to call when it's selected, this seems to be PopupMenu::activate_item(int p_idx) but it seems like this is assigned to p_callback and not p_key_callback in the various *add_item calls made from PopupMenu. So shouldn't it simply be 
```
if (value->callback != Callable()) { 
 	// If custom callback is set, use it. 
 	value->callback.call(value->meta); 
```
?

It also seems a little strange to me with the [[[NSApplication sharedApplication] keyWindow] sendEvent:event]; if there is no callback set. This feels like it could cause some sort of recursion since `- (BOOL)menuHasKeyEquivalent:(NSMenu *)menu forEvent:(NSEvent *)event target:(id *)target action:(SEL *)action` is likely called while processing events and the delegate method would I expect only return a matching target/action when appropriate and not need to initiate processing of non-matching events. It could be that this is what is causing the highlight of the item even though the action isn't triggered?

**Comment by lostminds** [comment]
I think this is resolved now with your recent menu bar improvements @bruvzg being merged? Or is there still something we're waiting for with this specific issue?

**Comment by akien-mga** [comment]
Fixed by #82900.
